src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs

```diff
@@ -17,178 +17,75 @@ public static partial class JsonSerializer
             {
                 // Re-throw with Path information.
-                ThrowHelper.ReThrowWithPath(readStack, ex);
+                ThrowHelper.ReThrowWithPath(state, ex);
             }
             catch (FormatException ex) when (ex.Source == ThrowHelper.ExceptionSourceValueToRethrowAsJsonException)
             {
-                ThrowHelper.ReThrowWithPath(readStack, reader, ex);
+                ThrowHelper.ReThrowWithPath(state, reader, ex);
             }
             catch (InvalidOperationException ex) when (ex.Source == ThrowHelper.ExceptionSourceValueToRethrowAsJsonException)
             {
-                ThrowHelper.ReThrowWithPath(readStack, reader, ex);
+                ThrowHelper.ReThrowWithPath(state, reader, ex);
             }
             catch (JsonException ex)
             {
-                ThrowHelper.AddExceptionInformation(readStack, reader, ex);
+                ThrowHelper.AddExceptionInformation(state, reader, ex);
                 throw;
             }
-
-            readStack.BytesConsumed += reader.BytesConsumed;
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static bool HandleObjectAsValue(
-            JsonTokenType tokenType,
+        internal static object? ReadCoreReEntry(
             JsonSerializerOptions options,
             ref Utf8JsonReader reader,
-            ref ReadStack readStack,
-            ref JsonReaderState initialState,
-            long initialBytesConsumed)
+            ref ReadStack state)
         {
-            if (readStack.ReadAhead)
-            {
-                // Attempt to skip to make sure we have all the data we need.
-                bool complete = reader.TrySkip();
-
-                // We need to restore the state in all cases as we need to be positioned back before
-                // the current token to either attempt to skip again or to actually read the value in
-                // HandleValue below.
-
-                reader = new Utf8JsonReader(
-                    reader.OriginalSpan.Slice(checked((int)initialBytesConsumed)),
-                    isFinalBlock: reader.IsFinalBlock,
-                    state: initialState);
-                Debug.Assert(reader.BytesConsumed == 0);
-                readStack.BytesConsumed += initialBytesConsumed;
-
-                if (!complete)
-                {
-                    // Couldn't read to the end of the object, exit out to get more data in the buffer.
-                    return false;
-                }
-
-                // Success, requeue the reader to the token for HandleValue.
-                reader.Read();
-                Debug.Assert(tokenType == reader.TokenType);
-            }
-
-            HandleValue(tokenType, options, ref reader, ref readStack);
-            return true;
+            JsonPropertyInfo jsonPropertyInfo = state.Current.JsonPropertyInfo!;
+            JsonConverter converter = jsonPropertyInfo!.ConverterBase;
```

Comment:
What's guaranteeing that `jsonPropertyInfo` isn't nullable. And if so, why do we need the `!` on line 85?

Reply:
We no longer need the ! on 85.

This method is currently only called in one flow-of-control and the jsonPropertyInfo is obtained in InitializeReEntry() and can't be null.